cpp/mrc/include/mrc/node/channel_holder.hpp

```diff
@@ -0,0 +1,1041 @@
+        edge->add_disconnector([this]() {
+            this->m_owned_edge_lifetime.reset();
+            this->m_owned_edge.reset();
+        });
+
+        // Set to the temp edge to ensure its alive until get_edge is called
+        m_owned_edge_lifetime = edge;
+
+        // Set to the weak ptr as well
+        m_owned_edge = edge;
+    }
+
+    void init_connected_edge(std::shared_ptr<EdgeHandle<T>> edge)
+    {
+        // Check for active connections
+        this->check_active_connection();
+
+        m_connected_edge = edge;
+    }
+
+    std::shared_ptr<EdgeHandleObj> get_edge_connection() const
+    {
+        if (auto edge = m_owned_edge.lock())
+        {
+            return std::shared_ptr<EdgeHandleObj>(new EdgeHandleObj(EdgeTypePair::create<T>(), edge));
+        }
+
+        throw std::runtime_error("Must set an edge before calling get_edge");
+    }
+
+    void make_edge_connection(std::shared_ptr<EdgeHandleObj> edge_obj)
+    {
+        CHECK(edge_obj->get_type() == EdgeTypePair::create<T>())
+            << "Incoming edge connection is not the correct type. Make sure to call "
+               "`EdgeBuilder::adapt_ingress<T>(edge)` or `EdgeBuilder::adapt_egress<T>(edge)` before calling "
+               "make_edge_connection";
+
+        // Unpack the edge, convert, and call the inner set_edge
+        auto unpacked_edge = edge_obj->get_handle_typed<EdgeHandle<T>>();
+
+        this->set_edge_handle(unpacked_edge);
+    }
+
+    void release_edge_connection()
+    {
+        m_owned_edge_lifetime.reset();
+        m_connected_edge.reset();
+    }
+
+    const std::shared_ptr<EdgeHandle<T>>& get_connected_edge() const
+    {
+        return m_connected_edge;
+    }
+
+  private:
+    void set_edge_handle(std::shared_ptr<EdgeHandle<T>> edge)
+    {
+        // Check for active connections
+        this->check_active_connection();
+
+        // Set to the temp edge to ensure its alive until get_edge is called
+        m_connected_edge = edge;
+
+        // Reset the weak_ptr since we dont own this edge
+        m_owned_edge.reset();
+
+        // Remove any init lifetime
+        m_owned_edge_lifetime.reset();
+
+        // Now indicate that we have a connection
+        edge->connect();
+    }
+
+    // Used for retrieving the current edge without altering its lifetime
+    std::weak_ptr<EdgeHandle<T>> m_owned_edge;
+
+    // This object ensures that any initialized edge is kept alive and is cleared on connection
+    std::shared_ptr<EdgeHandle<T>> m_owned_edge_lifetime;
+
+    // Holds a pointer to any set edge (different from init edge). Maintains lifetime
+    std::shared_ptr<EdgeHandle<T>> m_connected_edge;
+
+    // Allow edge builder to call set_edge
+    friend EdgeBuilder;
+
+    // Allow multi edge holder to access protected elements
+    template <typename KeyT, typename OtherT>
+    friend class MultiEdgeHolder;
+};
+
+template <typename KeyT, typename T>
+class MultiEdgeHolder
```

Comment:
How would that work? `MultiEdgeHolder` requires a key type. For a single `EdgeHolder` there is no key type. Not sure how this would work. Open to ideas

Reply:
I'm not entirely sure why KeyT needs to be a template parameter; but, what about something like:

```c++
template <typename KeyT=std::size_t, typename T>
class MultiEdgeHolder { ... }

template<typename T>
using EdgeHolder = MultiEdgeHolder<,T>
```

Reply:
Key is a template parameter because you can have different types of separation between the output nodes. For example, broadcast just uses size_t, but some may use strings for named outputs, and we have a test case that uses an enum. Having it be a template parameter supports all these cases.